### ISSUE TITLE
Client - Lagging of core (BB #802)

### ISSUE THREAD

**Issue by OregonAsari** [OregonCore/OregonCore#802]
This issue was migrated from bitbucket.
**Original Reporter:** smoldar
**Original Date:** 29.01.2015 22:32:32 GMT+0000
**Original Priority:** major
**Original Type:** bug
**Original State:** closed
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802

<hr>

Player tryes to login again, AccountId = XXX
Player tryes to login again, AccountId = XXX
Player tryes to login again, AccountId = XXX
Player tryes to login again, AccountId = XXX

100x spammed in server.log / sec i think about packet attack of core, when this spam runned, core was laggy 50-5000ms

I think this problem is here, this may be blocked by different way i think.

Destroying of client session, or something else to avoid laggy, because client can repeatly spam core.

CharacterHandler.cpp

```
#!c++

void WorldSession::HandlePlayerLoginOpcode(WorldPacket & recv_data)
{
    if (PlayerLoading() || GetPlayer() != NULL)
    {
        sLog.outError("Player tryes to login again, AccountId = %d",GetAccountId());
        return;
    }
```

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** smoldar
**Original Date:** 29.01.2015 22:33:13 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15340960

<hr>

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** smoldar
**Original Date:** 30.01.2015 10:05:41 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15349490

<hr>

On Trinitycore i found checking of packet size.

```
#!c++

    CHECK_PACKET_SIZE(recv_data,8);

    if(PlayerLoading() || GetPlayer() != NULL)
    {
        sLog.outError("Player tryes to login again, AccountId = %d",GetAccountId());
        return;
    }
```

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** desteny
**Original Date:** 30.01.2015 17:36:10 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15359134

<hr>

CHECK_PACKET_SIZE helps only if the packet is in wrong format, but i dont think so, the request for loggin is correct, and this scrit / bot / whatever try to hack the password by bruteforce or word lists! (i think)

there is a simple workaround, just allow 5 login per minute from the same IP (or 5 wrong logins) else dont even ask the database, just give error for x time.

this could be made with config option... will make sense.

!! KEEP IN MIND!! there are a lot of opcodes they are probably causing laggs by spamming requests to the server, this could be a problem in any case

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Alexdeath
**Original Date:** 31.01.2015 19:14:19 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15374585

<hr>

With 4 people online ping is 1500. Please, help with this....

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Alex_Step
**Original Date:** 31.01.2015 20:10:07 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15375028

<hr>

I confirm that there is an attack on the server using these packages. I would like to find a solution quickly :)

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** LordUsagi
**Original Date:** 31.01.2015 23:57:46 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15383573

<hr>

```
#!c++

    if (PlayerLoading() || GetPlayer() != NULL)
    {
        sLog.outError("Player tryes to login again, AccountId = %d",GetAccountId());
        KickPlayer();
        return;
    }
```

try

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** smoldar
**Original Date:** 01.02.2015 12:43:46 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15388090

<hr>

I meaned about kick or destroying of player session.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Eldro
**Original Date:** 01.02.2015 13:27:18 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15388456

<hr>

```
#!c++

    if (PlayerLoading() || GetPlayer() != NULL)
    {
        sLog.outError("Player tryes to login again, AccountId = %d",GetAccountId());
        KickPlayer();
        return;
    }
```

will not help

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** LordUsagi
**Original Date:** 01.02.2015 15:53:05 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15390056

<hr>

Then what will friend?

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Zaffy
**Original Date:** 01.02.2015 16:23:33 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15390389

<hr>

Won't help or doesn't help? I encourage you (all others) to test this first.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Alex_Step
**Original Date:** 01.02.2015 17:38:24 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15391177

<hr>

I can not say exactly helps or not, since I do not know how to reproduce this error. But I did not see her anymore.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Eldro
**Original Date:** 01.02.2015 19:17:30 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15392262

<hr>

It's very interesting.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** LordUsagi
**Original Date:** 01.02.2015 20:17:38 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15392945

<hr>

You use new code and then no more spam?

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** smoldar
**Original Date:** 01.02.2015 20:47:57 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15393288

<hr>

I dont know now, attacker which sended these fake packets get IP ban, now i dont know how to reproduce this attack on core.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** smoldar
**Original Date:** 01.02.2015 20:48:43 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15393301

<hr>

But when i will see this log again i will tell you that problem was fixed.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** smoldar
**Original Date:** 01.02.2015 20:50:41 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15393315

<hr>

I get info about attackers from russia deathside.ru they using ddos attacks and other attacks on realm and core, realm attacks i fixed by IP tables but when is logged on acc, packets are marked as right, but repeating.

I think this will fix problem, KickPlayer() destroy logged session of account. This will be blocked by this i mean.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Alex_Step
**Original Date:** 01.02.2015 20:50:45 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15393316

<hr>

A similar situation, I do not know exactly whether the attack stopped or helped the decision.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** desteny
**Original Date:** 02.02.2015 16:05:19 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15419141

<hr>

To Protect your Server from DDos Attacks you should use third party software of protection of your provider.

To code this in OC is not solving the problem at all, cause a massive ddos even is the core handles the best possible way will close your bandwidth your connections and finaly causing laggs and disconnects.

But if the attacks are not so big you can "work around" the problem.
1: more checks
2: maybe cache login results
3: implement counters for strange usage or the login procedure

but finaly the point, its all a workaround...

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Zaffy
**Original Date:** 02.02.2015 16:16:55 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15419466

<hr>

DoS attacks can be filtered easily in linux using iptables.  
But DDoS attacks are very subtle because they cannot be easily filtered.

Basically the difference is that DoS is coming from one IP - you can set limit on one ip, but DDoS is very similar but with that difference that there are many computers doing a DoS attack at the same time, and all seems to be legit.

Although I agree that this things can be filtered out outside of OregonCore with firewall rules, it would be nice to have better security in OC itself.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** smoldar
**Original Date:** 02.02.2015 16:33:00 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15419909

<hr>

I have a lot of things vs. ddos but this is not ddos attack, this attack seems as attack when client logged on account correctly, after from his own ip he attacked on my core, by repeatly sending packets (packets of log ingame)

I have some servers for normal bussiness ;) and on my servers iam under ddos attacks every day ;) 90% attacks between 300-2000mb/s it is not problem for me.

I have iptables filters, but this was legitimated packets, but repeated log-in.

But... when ddos attack overflood your connectivity, iptables wont help :-) PC will be protected, but network not :-)

But on my business i have migrating policies of these attacks.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** smoldar
**Original Date:** 02.02.2015 16:35:08 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15419973

<hr>

But when you are under attacks by BOTNETs... you are died :D if you have VPS .. they will spam your network more than your provider allows for you ;) and your project will be laggy... By iptables you will only defend your pc ( in last year one of my servers crashed... RJ45 fired :O by massive ddos attack) I never saw fired RJ45 connector :-)

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** desteny
**Original Date:** 03.02.2015 17:02:46 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15448347

<hr>

but if you have 100x spammed in logs per sec (from the same user!!) then this is not a normal client action, for sure a attack.
but if this is from the same client, then its easy to block.

just create a map (int, int) in c++ using accountid as key and count as value.
then define a threshold like 5 or 10 login actions alowed in 1 min (no normal player will log and relog as often)

to keep it simple if the count is > then your threshold kick the session or just do nothing (silent return)

every minute you clear the map (or define your interval)

this simple mod will reduce the server load created by login spam significantly and has no impact on normal player

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Zaffy
**Original Date:** 03.02.2015 17:49:49 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15449381

<hr>

desteny that solution would work with only this specific situation.

I think better would be to either to check the same but with sent packets, or process only about 15-30 packets in an update cycle for each session. Additionally, this could be bypassed by connecting multiple accounts, so I would also set a limit for accounts online for one ip.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** smoldar
**Original Date:** 03.02.2015 19:40:16 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15451658

<hr>

Desteny, it is not easy to block ;)

I blocked IP of client and his accounts... But it is not prevent of problem, you can try block me permanently.. and i can 1000x change ip/emails and again send these fake packets.. it is not solve and packets are marked as right...

I have limit 20 connections / ip .. but this is not count of connections, only repeated packets of one connection. But using of proxy, this attack will be lowered... but i reported it because this cannot be blocked by iptables ;) i have a lot of rules which limiting connections

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** LordUsagi
**Original Date:** 03.02.2015 19:47:30 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15451796

<hr>

Is it still happening or not?

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** smoldar
**Original Date:** 03.02.2015 19:52:19 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15451899

<hr>

i dont know now...

I blocked attacker's ip(in firewall) and accounts, for avoid of this problem again, he must use proxy or friends network for again attack. This script is applied for last 2days and i never see again this problem in logs. If i will see this problem again, i will report it.

I think now can be marked as resolved

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Alex_Step
**Original Date:** 03.02.2015 20:17:46 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15452389

<hr>

After applying this solution - I have not seen more in the log of this error. But the attacker still create new accounts, and players complained about the lag on the server. So I think the error has simply ceased to appear in the logs, but it has not solved the problem itself.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Zaffy
**Original Date:** 03.02.2015 20:38:27 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15452757

<hr>

@Alex_Step Try this and let me know how it goes - it limits packets processed in one cycle to 30. You can also adjust it by your needs.

```
diff --git a/src/game/WorldSession.cpp b/src/game/WorldSession.cpp
--- a/src/game/WorldSession.cpp
+++ b/src/game/WorldSession.cpp
@@ -183,7 +183,8 @@
     // Retrieve packets from the receive queue and call the appropriate handlers
     // not proccess packets if socket already closed
     WorldPacket* packet;
-    while (m_Socket && !m_Socket->IsClosed() && _recvQueue.next(packet))
+    uint32 i = 0;
+    while (m_Socket && !m_Socket->IsClosed() && ++i <= 30 && _recvQueue.next(packet))
     {
         /*#if 1
         sLog.outError("MOEP: %s (0x%.4X)",
```

If the problem reappears we will need to lower the limit and possibly, disconnect the player if there are many packets queued. (memory consuming - allocating might be also slowing down the core)

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** LordUsagi
**Original Date:** 04.02.2015 04:20:32 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15459263

<hr>

Should check out 59b4c3492482b7536651625f0b1ed0ef87f52e79 TrinityCore

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** ShanduLIVE
**Original Date:** 04.02.2015 08:08:21 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15462081

<hr>

 [https://github.com/TrinityCore/TrinityCore/commit/59b4c3492482b7536651625f0b1ed0ef87f52e79](https://github.com/TrinityCore/TrinityCore/commit/59b4c3492482b7536651625f0b1ed0ef87f52e79)

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** smoldar
**Original Date:** 04.02.2015 08:10:33 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15462115

<hr>

Thanks, i remember this. I saw these config options in some trinity cores on higher patches :)

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** ShanduLIVE
**Original Date:** 04.02.2015 08:25:46 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15462718

<hr>

You're welcome.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** desteny
**Original Date:** 04.02.2015 15:41:08 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15473148

<hr>

@smoldar you told uppon "100x spammed in logs per sec (from the same user!!)"

even if the same user (from multiple IP) you can limit by user
if you have a real ddos situation the user and ip changes then it could be catched this way

@zaffy its true this is only a solution for this particulary problem, it doesnt work on ddos with different accounts

about the limiting te queue process to 30 packages per cycle could help but its not a nice solution and can have strange effects.

btw. if the session (user) changes even then it could be  x \* 30 actions per cycle.

from this point of view its the same as limiting the login action only (per session / account) only that it will have no effect on other opcodes in queue

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Zaffy
**Original Date:** 04.02.2015 15:54:17 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15473454

<hr>

@desteny, what are the strange effects? Keep in mind that you're just executing 30 opcodes per cycle (50ms) but you don't limit the queue - there will be still all packets in ordered order as they came with the difference they won't be executed at once.

When I did talk about it would work only in this case I meant attacker could spam any opcode, not just this one.

The limiting per cycle is very useful and common not just for security - for example warcraft III pathfinding doesn't calculate path for all units in one cycle but rather balances those to multiple cycles, so you as a player won't notice a lag.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** smoldar
**Original Date:** 04.02.2015 16:52:23 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15475446

<hr>

Desteny.. It was normal packet attack from 1 client ! Not ddos attack ...

One user repeatly sended packet about log of character into game...

Packets was rightful... but sended repeatly and core was lagged by this, because core repeatly loaded character.

I know how block ddos and other limits.. i have limits of login on account / sec.. .but client was logged only 1 time... and sended 50x same packet for log into game nothing else. It was attack on core.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** desteny
**Original Date:** 05.02.2015 17:24:53 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15502173

<hr>

@Zaffy, ok you may right in this case, the limitations can benefit the system, and from the point of "strange effects" its maybe my fault, cause i was thinking about timing problems between sessions (players) cause of queueing commands... but in real life you will never hit the limit in 50 ms (that is true)

but the thing i wanted to point out is that 30 login tryes per cycle are 600 per second, and thats to much, even if we have ony one session!

even if you limit the count to 1 are still 20 per second possible per session (with multiple sessions the problem will get increased)...

for this reason i suggested a longer period of time like allowed logins per minute!

@smoldar " It was normal packet attack from 1 client" => OK
"One user repeatly sended packet about log of character into game...
Packets was rightful... but sended repeatly and core was lagged by this, because core repeatly loaded character."

a normal client will at max do 1 login each 10 seconds! even on fast PC, then it could never do 100x per second!

there is a "client"  / "bot" / "whatever" who send correcct commands (like a client)
for example you can use the mangos client for 2.4.3 for login part, it works, you code a cpp loop and you have your spam client!

repeatly loading character with "WorldSession::HandlePlayerLoginOpcode"??
this opcode is only for login!

in this case you may have the same problem on other opcodes like WHO list or Ticket! they are possible to exploit!

its like i told, each opcode needs his own strategy how to handle it (my opinion) maybe by cache, by limits or checks of ingame status (like quest dialog only shown if target = questgiver npc)

" i have limits of login on account / sec.. .but client was logged only 1 time... and sended 50x same packet for log into game nothing else. It was attack on core."

yes, the login it self (on realm) is limited, but not the WorldSession::HandlePlayerLoginOpcode! this is called if you login to the world, not into realm! it needs his own limitations!

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** desteny
**Original Date:** 05.02.2015 17:31:37 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15502331

<hr>

Note on the end: @Zaffy 
as you know the WorldSession::HandlePlayerLoginOpcode has not much stuff to do, if you spam it even 100 times its fast enought to process.

but at the end it calles a DelayQueryHolder with massive work on the callback!

this will result in a spam of async actions with much processing (and may DB) usage, this are blocking hardware resources and this causes laggs.

for this reason even a limited execution per update cycle will cause a long queue of commands with a loot of stress for the machine...

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** smoldar
**Original Date:** 05.02.2015 17:32:54 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15502368

<hr>

It is not log on account, he logged only 1x on acc but sended XX count of requests for log char into game only this.

I have limited logins on realm(i had realm under attacks 50x login /sec with multiple ips) but i blocked it by IPTABLES max 2 logins / 3sec from ip... but this is only spam of joining into game

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** smoldar
**Original Date:** 05.02.2015 17:41:38 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15502538

<hr>

I think now can be changed as resolved, i will report or reopen it if i see any other these attacks.

Applied last commits with limit of packets from Zaffy and Kick commit from LordUsagi (thank to all)

I never seen this attack again, because attacker get block of his own ip with accounts.

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** LordUsagi
**Original Date:** 05.02.2015 20:01:23 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15505353

<hr>

I think we should keep this open for 2 more days, considering you're a guinea pig with the packet limiter and whatever you might experience before we apply :)

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** smoldar
**Original Date:** 05.02.2015 20:20:17 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15505763

<hr>

Iam tester :-D

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** LordUsagi
**Original Date:** 05.02.2015 22:03:06 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15507521

<hr>

xD yeah!

**Comment by OregonAsari** [comment]
This comment was migrated from bitbucket.
**Original User:** Zaffy
**Original Date:** 11.02.2015 16:00:01 GMT+0000
**Direct Link:** https://bitbucket.org/oregon/oregoncore/issues/802/#comment-15626637

<hr>
